// @softarc/sheriff-core/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProjectData = exports.UserError = exports.noDependencies = exports.sameTag = exports.anyTag = exports.violatesDependencyRule = exports.hasDeepImport = void 0;
var deep_import_1 = require("./lib/eslint/deep-import");
Object.defineProperty(exports, "hasDeepImport", { enumerable: true, get: function () { return deep_import_1.hasDeepImport; } });
var violates_dependency_rule_1 = require("./lib/eslint/violates-dependency-rule");
Object.defineProperty(exports, "violatesDependencyRule", { enumerable: true, get: function () { return violates_dependency_rule_1.violatesDependencyRule; } });
var any_tag_1 = require("./lib/checks/any-tag");
Object.defineProperty(exports, "anyTag", { enumerable: true, get: function () { return any_tag_1.anyTag; } });
var same_tag_1 = require("./lib/checks/same-tag");
Object.defineProperty(exports, "sameTag", { enumerable: true, get: function () { return same_tag_1.sameTag; } });
var no_dependencies_1 = require("./lib/checks/no-dependencies");
Object.defineProperty(exports, "noDependencies", { enumerable: true, get: function () { return no_dependencies_1.noDependencies; } });
var user_error_1 = require("./lib/error/user-error");
Object.defineProperty(exports, "UserError", { enumerable: true, get: function () { return user_error_1.UserError; } });
var get_project_data_1 = require("./lib/api/get-project-data");
Object.defineProperty(exports, "getProjectData", { enumerable: true, get: function () { return get_project_data_1.getProjectData; } });
