// @softarc/sheriff-core/src/lib/api/get-project-data.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProjectData = getProjectData;
const fs_path_1 = require("../file-info/fs-path");
const init_1 = require("../main/init");
const calc_tags_for_module_1 = require("../tags/calc-tags-for-module");
const traverse_file_info_1 = require("../modules/traverse-file-info");
const getFs_1 = __importDefault(require("../fs/getFs"));
function calcOrGetTags(modulePath, projectInfo, tagsCache) {
    if (modulePath in tagsCache) {
        return tagsCache[modulePath];
    }
    const tags = (0, calc_tags_for_module_1.calcTagsForModule)(modulePath, projectInfo.rootDir, projectInfo.config.tagging, projectInfo.config.autoTagging);
    tagsCache[modulePath] = tags;
    return tags;
}
function getProjectData(entryFile, cwd) {
    const fs = (0, getFs_1.default)();
    const absoluteEntryFile = cwd === undefined ? entryFile : fs.join(cwd, entryFile);
    const projectInfo = (0, init_1.init)((0, fs_path_1.toFsPath)(absoluteEntryFile));
    const data = {};
    const tagsCache = {};
    for (const { fileInfo } of (0, traverse_file_info_1.traverseFileInfo)(projectInfo.fileInfo)) {
        data[fileInfo.path] = {
            module: fileInfo.moduleInfo.directory || '.',
            tags: calcOrGetTags(fileInfo.moduleInfo.directory, projectInfo, tagsCache),
            imports: fileInfo.imports.map((fileInfo) => fileInfo.path),
        };
    }
    return relativizeIfRequired(data, cwd);
}
function relativizeIfRequired(data, cwd) {
    if (cwd === undefined) {
        return data;
    }
    const fs = (0, getFs_1.default)();
    const relative = (path) => fs.relativeTo(cwd, path) || '.';
    const relativizedData = {};
    for (const [modulePath, moduleData] of Object.entries(data)) {
        relativizedData[relative((0, fs_path_1.toFsPath)(modulePath))] = {
            module: relative((0, fs_path_1.toFsPath)(moduleData.module)),
            tags: moduleData.tags,
            imports: moduleData.imports.map((importPath) => relative((0, fs_path_1.toFsPath)(importPath))),
        };
    }
    return relativizedData;
}
